kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs

```diff
@@ -0,0 +1,420 @@
+
+        // Run the micro-passes that clean up bodies.
+        for pass in charon_lib::transform::ULLBC_PASSES.iter() {
+            pass.transform_ctx(&mut ccx)
+        }
+
+        // # Go from ULLBC to LLBC (Low-Level Borrow Calculus) by reconstructing
+        // the control flow.
+        charon_lib::ullbc_to_llbc::translate_functions(&mut ccx);
+
+        trace!("# LLBC resulting from control-flow reconstruction:\n\n{}\n", ccx);
+
+        // Run the micro-passes that clean up bodies.
+        for pass in charon_lib::transform::LLBC_PASSES.iter() {
+            pass.transform_ctx(&mut ccx)
+        }
+
+        // Print the LLBC if requested. This is useful for expected tests.
+        if queries.args().print_llbc {
+            println!("# Final LLBC before serialization:\n\n{}\n", ccx);
+        } else {
+            debug!("# Final LLBC before serialization:\n\n{}\n", ccx);
+        }
+
+        // Display an error report about the external dependencies, if necessary
+        ccx.errors.report_external_deps_errors();
+
+        let crate_data: charon_lib::export::CrateData = charon_lib::export::CrateData::new(&ccx);
+
+        // No output should be generated if user selected no_codegen.
+        if !tcx.sess.opts.unstable_opts.no_codegen && tcx.sess.opts.output_types.should_codegen() {
+            // # Final step: generate the files.
+            // `crate_data` is set by our callbacks when there is no fatal error.
+            let mut pb = llbc_file.to_path_buf();
+            pb.set_extension("llbc");
+            println!("Writing LLBC file to {}", pb.display());
+            if let Err(()) = crate_data.serialize_to_file(&pb) {
+                tcx.sess.dcx().err("Failed to write LLBC file");
+            }
+        }
+
+        (items, None)
+    }
+}
+
+impl CodegenBackend for LlbcCodegenBackend {
+    fn metadata_loader(&self) -> Box<MetadataLoaderDyn> {
+        Box::new(rustc_codegen_ssa::back::metadata::DefaultMetadataLoader)
+    }
+
+    fn provide(&self, providers: &mut Providers) {
+        provide::provide(providers, &self.queries.lock().unwrap());
+    }
+
+    fn print_version(&self) {
+        println!("Kani-llbc version: {}", env!("CARGO_PKG_VERSION"));
+    }
+
+    fn locale_resource(&self) -> &'static str {
+        // We don't currently support multiple languages.
+        DEFAULT_LOCALE_RESOURCE
+    }
+
+    fn codegen_crate(
+        &self,
+        tcx: TyCtxt,
+        rustc_metadata: EncodedMetadata,
+        _need_metadata_module: bool,
+    ) -> Box<dyn Any> {
+        let ret_val = rustc_internal::run(tcx, || {
+            // Queries shouldn't change today once codegen starts.
+            let queries = self.queries.lock().unwrap().clone();
+
+            // Codegen all items that need to be processed according to the selected reachability mode:
+            //
+            // - Harnesses: Generate one model per local harnesses (marked with `kani::proof` attribute).
```

Comment:
I'm curious. Do we want to keep this model for Aeneas backend too? Not something that we need to change for this PR, but it's something that it may be worth thinking about.

Reply:
Good point. For Aeneas, I was thinking that the public function mode would be the default.